staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -1500,3 +1433,322 @@ func TestGetRequestWaitContext(t *testing.T) {
 		})
 	}
 }
+
+func TestPriorityAndFairnessWithRequestDeadline(t *testing.T) {
+	// expectations are captured into reusable instances
+	var (
+		wantStreamResetErr            = wantContains{"stream error: stream ID 1; INTERNAL_ERROR; received from peer"}
+		wantHTTPStatusOK              = wantStatusCode{http.StatusOK}
+		wantHTTPStatusGatewayTimeout  = wantStatusCode{http.StatusGatewayTimeout}
+		wantHTTPStatusTooManyRequests = wantStatusCode{http.StatusTooManyRequests}
+	)
+
+	// group names are added here
+	const (
+		withTimeoutFilter = "WithTimeoutFilter"
+	)
+
+	// all sub-tests must adhere to the following
+	type tRunner interface {
+		runT(t *testing.T)
+	}
+
+	// the following sub-tests are intended to be run in parallel
+	// under a certain feature gate configuration
+	tests := []struct {
+		name    string
+		runners map[string]tRunner
+	}{
+		{
+			name: "PanicShouldNotRejectFutureRequest",
+			runners: map[string]tRunner{
+				withTimeoutFilter: panicShouldNotRejectFutureRequest{
+					firstReqVerifier: wantBoth{
+						a: wantStreamResetErr,
+						b: clientMustNotTimeout{},
+					},
+					secondReqVerifier: wantBoth{
+						a: wantHTTPStatusOK,
+						b: wantAPFResponseHeaders{},
+					},
+				},
+			},
+		},
+		{
+			name: "RequestTimesOutBeforeHandlerWrites",
+			runners: map[string]tRunner{
+				withTimeoutFilter: requestTimesOutBeforeHandlerWrites{
+					reqVerifier: wantBoth{
+						a: wantHTTPStatusGatewayTimeout,
+						b: clientMustNotTimeout{},
+					},
+				},
+			},
+		},
+		{
+			name: "HandlerPanicsAfterRequestTimesOut",
+			runners: map[string]tRunner{
+				withTimeoutFilter: handlerPanicsAfterRequestTimesOut{
+					reqVerifier: wantBoth{
+						a: wantHTTPStatusGatewayTimeout,
+						b: clientMustNotTimeout{},
+					},
+				},
+			},
+		},
+		{
+			name: "HandlerWritesBeforeRequestTimesOut",
+			runners: map[string]tRunner{
+				withTimeoutFilter: handlerWritesBeforeRequestTimesOut{
+					reqVerifier: wantBoth{
+						a: wantStreamResetErr,
+						b: clientMustNotTimeout{},
+					},
+				},
+			},
+		},
+		{
+			name: "EnqueuedRequestTimingOut",
+			runners: map[string]tRunner{
+				withTimeoutFilter: enqueuedRequestTimingOut{
+					firstReqVerifier: wantBoth{
+						a: wantHTTPStatusGatewayTimeout,
+						b: clientMustNotTimeout{},
+					},
+					secondReqVerifier: wantBoth{
+						a: wantHTTPStatusTooManyRequests,
+						b: wantBoth{
+							a: clientMustNotTimeout{},
+							b: wantAPFResponseHeaders{},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	// each the following parent tests represent a group, and must
+	// run serially to each other
+	groups := []struct {
```

Comment:
Since there is actually only one group, this looks more complicated than necessary.
Is there an expectation of more groups in the future? If so then maybe a comment about that would be helpful. If not, then I think simplifying the test code would be in order.

Reply:
Yes, we will have at least two groups, with the proposed feature enabled/disabled. I added a comment